<a href="https://colab.research.google.com/github/Seb85vickz/CMP7005-Programming-for-Data-Analysis/blob/main/SEB_Week5_workshop4__Pandas_Part2_(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **🧭 Task 1 – Fundamental Data Understanding (Indian Air Pollution Data)**

### **🎯 Objective :**

The purpose of this task is to help you explore and understand the structure and quality of the Indian Air Pollution Dataset provided for your assessment.
You will use Pandas to:

* Combine multiple CSV files into a single dataset, and

* Perform fundamental data understanding (data inspection, summary statistics, and missing value analysis).

## **Mounting the drive**

In [ ]:
# Load the Drive helper and mount
from google.colab import drive

# This will prompt for authorization.
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
%cd '/content/drive/MyDrive/Colab/CMP7005'
# please change the path according to the location of your data

/content/drive/MyDrive/Colab/CMP7005


In [ ]:
%ls # it shows all the content of your folder if done properly

Ahmedabad_data.csv       Chennai_data.csv     Kochi_data.csv
Aizawl_data.csv          Coimbatore_data.csv  Kolkata_data.csv
all_cities_combined.csv  Delhi_data.csv       Lucknow_data.csv
Amaravati_data.csv       Ernakulam_data.csv   Mumbai_data.csv
Amritsar_data.csv        Gurugram_data.csv    Patna_data.csv
Bengaluru_data.csv       Guwahati_data.csv    Shillong_data.csv
Bhopal_data.csv          Hyderabad_data.csv   Talcher_data.csv
Brajrajnagar_data.csv    Jaipur_data.csv      Thiruvananthapuram_data.csv
Chandigarh_data.csv      Jorapokhar_data.csv  Visakhapatnam_data.csv


# **🧩 What You Need to Do:**

**Step 1: Import Libraries :**

Start by importing the necessary Python libraries:

In [ ]:
import pandas as pd
import os

### **🧩 Merging of the csv files:**



### **Combining all the csv files in the drive path: You can use either of the merging options:**  
---

**🧩 OPTION 1 :**

In [ ]:
# Import the necessary libraries
import pandas as pd  # pandas is used for working with data tables
import glob         # glob is used to find files by name patterns

# STEP 1: The pattern "*_data.csv" means "find all files that end with '_data.csv'"
# This will find files like: Ahmedabad_data.csv, Delhi_data.csv, Mumbai_data.csv, etc.
city_files = glob.glob("*_data.csv")

# STEP 2: Create an empty list to store all our city data
# We'll put each city's data in this list before combining them
all_cities_data = []

# STEP 3: Read each city file one by one
for file_name in city_files:
    # Read the current city's CSV file into a DataFrame
    # A DataFrame is like a spreadsheet table in Python
    city_df = pd.read_csv(file_name)

    # Add this city's data to our list
    all_cities_data.append(city_df)

    # Optional: Print which file we just read
    print(f"Loaded: {file_name}")

# STEP 4: Combine all city data into one big table
# pd.concat() joins all the DataFrames in our list together
# ignore_index=True makes sure the row numbers are continuous (0, 1, 2, 3...)
combined_data = pd.concat(all_cities_data, ignore_index=True)

# STEP 5: Save the combined data to a new CSV file
# index=False means don't save the row numbers as a separate column
combined_data.to_csv("all_cities_combined.csv", index=False)

# STEP 6: Show us what we accomplished
# len(city_files) = count of how many city files we combined
# len(combined_data) = total number of rows in the final combined file
print(f"SUCCESS: Combined {len(city_files)} city files into one file with {len(combined_data)} total rows")
print("The combined file is saved as: all_cities_combined.csv")

Loaded: Aizawl_data.csv
Loaded: Gurugram_data.csv
Loaded: Brajrajnagar_data.csv
Loaded: Amritsar_data.csv
Loaded: Kochi_data.csv
Loaded: Patna_data.csv
Loaded: Delhi_data.csv
Loaded: Jorapokhar_data.csv
Loaded: Mumbai_data.csv
Loaded: Lucknow_data.csv
Loaded: Guwahati_data.csv
Loaded: Ernakulam_data.csv
Loaded: Shillong_data.csv
Loaded: Hyderabad_data.csv
Loaded: Coimbatore_data.csv
Loaded: Ahmedabad_data.csv
Loaded: Bhopal_data.csv
Loaded: Amaravati_data.csv
Loaded: Jaipur_data.csv
Loaded: Chennai_data.csv
Loaded: Kolkata_data.csv
Loaded: Bengaluru_data.csv
Loaded: Chandigarh_data.csv
Loaded: Visakhapatnam_data.csv
Loaded: Thiruvananthapuram_data.csv
Loaded: Talcher_data.csv
SUCCESS: Combined 26 city files into one file with 29531 total rows
The combined file is saved as: all_cities_combined.csv


In [ ]:
df= pd.read_csv('all_cities_combined.csv')
df

City        Date  PM2.5   PM10     NO    NO2    NOx    NH3    CO  \
0       Aizawl  11/03/2020  32.69  47.91   6.99   2.85  11.93  26.64  0.60   
1       Aizawl  12/03/2020  31.21  38.66   7.20   1.27  10.65  25.63  0.56   
2       Aizawl  13/03/2020  38.39  46.68   7.19   0.91  10.37  29.16  0.57   
3       Aizawl  14/03/2020  43.23  50.83   7.14   1.07  10.48  28.95  0.57   
4       Aizawl  15/03/2020  33.82  41.03   7.09   0.36   9.73  28.41  0.48   
...        ...         ...    ...    ...    ...    ...    ...    ...   ...   
29526  Talcher  27/06/2020  20.75  25.94   8.17  21.95  15.89   6.73  1.18   
29527  Talcher  28/06/2020  24.88  31.07    NaN    NaN   0.00    NaN  1.55   
29528  Talcher  29/06/2020  20.37  25.46  22.82  19.64  18.96  10.23  1.59   
29529  Talcher  30/06/2020  18.62  23.26  21.24  26.53  47.60   4.45  1.30   
29530  Talcher  01/07/2020  24.85  31.06  27.73  29.36  57.09   1.37  1.52   

         SO2     O3  Benzene  Toluene  Xylene    AQI    AQI_Bucket  
0       4.53   4.48     0.03     0.30     NaN    NaN           NaN  
1       4.22   2.81     0.01     0.08     NaN   52.0  Satisfactory  
2       4.46   0.18     0.00     0.00     NaN   60.0  Satisfactory  
3       4.53   0.41     0.00     0.00     NaN   62.0  Satisfactory  
4       4.63   0.30     0.00     0.00     NaN   70.0  Satisfactory  
...      ...    ...      ...      ...     ...    ...           ...  
29526  18.62  48.87     0.00      NaN     NaN   90.0  Satisfactory  
29527  21.33  48.55     0.00      NaN     NaN   98.0  Satisfactory  
29528  20.79  46.84     0.00      NaN     NaN   99.0  Satisfactory  
29529  14.34  28.77     0.00      NaN     NaN   86.0  Satisfactory  
29530  14.42  14.90     0.00      NaN     NaN  101.0      Moderate  

[29531 rows x 16 columns]

**🧩 OPTION 2 :**


The process involves iterating through all files in a specified directory (drive_path) to identify those with a .csv extension. Each identified CSV file is then read into a Pandas DataFrame, which is subsequently appended to a list (dataframes). Once all the CSV files are processed, the individual DataFrames in the list are combined into a single consolidated DataFrame (all_data) using the pd.concat() function. This consolidation ensures that the combined data is reindexed, creating a unified dataset for further analysis.

In [ ]:
drive_path = '/content/drive/MyDrive/Semester1_25/Beijing air quality'

In [ ]:
dataframes = []
for filename in os.listdir(drive_path):
    if filename.endswith('.csv'):  # Check if the file is a CSV file
        file_path = os.path.join(drive_path, filename)
        df = pd.read_csv(file_path)  # Read the CSV file into a DataFrame
        dataframes.append(df)  # Add the DataFrame to the list

In [ ]:
df1 = pd.concat(dataframes, ignore_index=True)

## **🧩 Perform Fundamental Data Understanding :**

**Once you have the merged dataset, explore and understand its structure.**

In [ ]:
df.head()

City        Date  PM2.5   PM10    NO   NO2    NOx    NH3    CO   SO2  \
0  Aizawl  11/03/2020  32.69  47.91  6.99  2.85  11.93  26.64  0.60  4.53   
1  Aizawl  12/03/2020  31.21  38.66  7.20  1.27  10.65  25.63  0.56  4.22   
2  Aizawl  13/03/2020  38.39  46.68  7.19  0.91  10.37  29.16  0.57  4.46   
3  Aizawl  14/03/2020  43.23  50.83  7.14  1.07  10.48  28.95  0.57  4.53   
4  Aizawl  15/03/2020  33.82  41.03  7.09  0.36   9.73  28.41  0.48  4.63   

     O3  Benzene  Toluene  Xylene   AQI    AQI_Bucket  
0  4.48     0.03     0.30     NaN   NaN           NaN  
1  2.81     0.01     0.08     NaN  52.0  Satisfactory  
2  0.18     0.00     0.00     NaN  60.0  Satisfactory  
3  0.41     0.00     0.00     NaN  62.0  Satisfactory  
4  0.30     0.00     0.00     NaN  70.0  Satisfactory

In [ ]:
df.tail()

City        Date  PM2.5   PM10     NO    NO2    NOx    NH3    CO  \
29526  Talcher  27/06/2020  20.75  25.94   8.17  21.95  15.89   6.73  1.18   
29527  Talcher  28/06/2020  24.88  31.07    NaN    NaN   0.00    NaN  1.55   
29528  Talcher  29/06/2020  20.37  25.46  22.82  19.64  18.96  10.23  1.59   
29529  Talcher  30/06/2020  18.62  23.26  21.24  26.53  47.60   4.45  1.30   
29530  Talcher  01/07/2020  24.85  31.06  27.73  29.36  57.09   1.37  1.52   

         SO2     O3  Benzene  Toluene  Xylene    AQI    AQI_Bucket  
29526  18.62  48.87      0.0      NaN     NaN   90.0  Satisfactory  
29527  21.33  48.55      0.0      NaN     NaN   98.0  Satisfactory  
29528  20.79  46.84      0.0      NaN     NaN   99.0  Satisfactory  
29529  14.34  28.77      0.0      NaN     NaN   86.0  Satisfactory  
29530  14.42  14.90      0.0      NaN     NaN  101.0      Moderate

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 29531 entries, 0 to 29530
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   City        29531 non-null  object 
 1   Date        29531 non-null  object 
 2   PM2.5       24933 non-null  float64
 3   PM10        18391 non-null  float64
 4   NO          25949 non-null  float64
 5   NO2         25946 non-null  float64
 6   NOx         25346 non-null  float64
 7   NH3         19203 non-null  float64
 8   CO          27472 non-null  float64
 9   SO2         25677 non-null  float64
 10  O3          25509 non-null  float64
 11  Benzene     23908 non-null  float64
 12  Toluene     21490 non-null  float64
 13  Xylene      11422 non-null  float64
 14  AQI         24850 non-null  float64
 15  AQI_Bucket  24850 non-null  object 
dtypes: float64(13), object(3)
memory usage: 3.6+ MB


In [ ]:
df.dtypes

City           object
Date           object
PM2.5         float64
PM10          float64
NO            float64
NO2           float64
NOx           float64
NH3           float64
CO            float64
SO2           float64
O3            float64
Benzene       float64
Toluene       float64
Xylene        float64
AQI           float64
AQI_Bucket     object
dtype: object

In [ ]:
df.columns

Index(['City', 'Date', 'PM2.5', 'PM10', 'NO', 'NO2', 'NOx', 'NH3', 'CO', 'SO2',
       'O3', 'Benzene', 'Toluene', 'Xylene', 'AQI', 'AQI_Bucket'],
      dtype='object')

In [ ]:
df.shape
print(f'No of Rows: {df.shape[0]}, No of Columns: {df.shape[1]}')

No of Rows: 29531, No of Columns: 16


### **Total number of stations in the dataset:**

In [ ]:
city_files = df['City'].value_counts()
print(f'Total number of Cities in the dataset : {len(city_files)}')
city_files


Total number of Cities in the dataset : 26


City
Lucknow               2009
Mumbai                2009
Delhi                 2009
Bengaluru             2009
Chennai               2009
Ahmedabad             2009
Hyderabad             2006
Patna                 1858
Gurugram              1679
Visakhapatnam         1462
Amritsar              1221
Jorapokhar            1169
Jaipur                1114
Thiruvananthapuram    1112
Amaravati              951
Brajrajnagar           938
Talcher                925
Kolkata                814
Guwahati               502
Coimbatore             386
Shillong               310
Chandigarh             304
Bhopal                 289
Kochi                  162
Ernakulam              162
Aizawl                 113
Name: count, dtype: int64

### **Displaying the percentage of missing value**

In [ ]:
def missing_values_table(df):
    # Total missing values
    mis_val = df.isnull().sum()

    # Percentage of missing values
    mis_val_percent = 100 * mis_val / len(df)

    mis_val_table = pd.concat([mis_val, mis_val_percent], axis=1)

    mis_val_table = mis_val_table.rename(columns={0: 'Missing Values', 1: '% of Total Values'})

    # Sort the table by percentage of missing descending
    mis_val_table = mis_val_table.sort_values('% of Total Values', ascending=False)

    return mis_val_table

missing_values = missing_values_table(df)
display(missing_values.style.background_gradient(cmap='Blues'))

In [ ]:
df.describe()

PM2.5          PM10            NO           NO2           NOx  \
count  24933.000000  18391.000000  25949.000000  25946.000000  25346.000000   
mean      67.450578    118.127103     17.574730     28.560659     32.309123   
std       64.661449     90.605110     22.785846     24.474746     31.646011   
min        0.040000      0.010000      0.020000      0.010000      0.000000   
25%       28.820000     56.255000      5.630000     11.750000     12.820000   
50%       48.570000     95.680000      9.890000     21.690000     23.520000   
75%       80.590000    149.745000     19.950000     37.620000     40.127500   
max      949.990000   1000.000000    390.680000    362.210000    467.630000   

                NH3            CO           SO2            O3       Benzene  \
count  19203.000000  27472.000000  25677.000000  25509.000000  23908.000000   
mean      23.483476      2.248598     14.531977     34.491430      3.280840   
std       25.684275      6.962884     18.133775     21.694928     15.811136   
min        0.010000      0.000000      0.010000      0.010000      0.000000   
25%        8.580000      0.510000      5.670000     18.860000      0.120000   
50%       15.850000      0.890000      9.160000     30.840000      1.070000   
75%       30.020000      1.450000     15.220000     45.570000      3.080000   
max      352.890000    175.810000    193.860000    257.730000    455.030000   

            Toluene        Xylene           AQI  
count  21490.000000  11422.000000  24850.000000  
mean       8.700972      3.070128    166.463581  
std       19.969164      6.323247    140.696585  
min        0.000000      0.000000     13.000000  
25%        0.600000      0.140000     81.000000  
50%        2.970000      0.980000    118.000000  
75%        9.150000      3.350000    208.000000  
max      454.850000    170.370000   2049.000000

In [ ]:
print(df.describe())

              PM2.5          PM10            NO           NO2           NOx  \
count  24933.000000  18391.000000  25949.000000  25946.000000  25346.000000   
mean      67.450578    118.127103     17.574730     28.560659     32.309123   
std       64.661449     90.605110     22.785846     24.474746     31.646011   
min        0.040000      0.010000      0.020000      0.010000      0.000000   
25%       28.820000     56.255000      5.630000     11.750000     12.820000   
50%       48.570000     95.680000      9.890000     21.690000     23.520000   
75%       80.590000    149.745000     19.950000     37.620000     40.127500   
max      949.990000   1000.000000    390.680000    362.210000    467.630000   

                NH3            CO           SO2            O3       Benzene  \
count  19203.000000  27472.000000  25677.000000  25509.000000  23908.000000   
mean      23.483476      2.248598     14.531977     34.491430      3.280840   
std       25.684275      6.962884     18.133775    

# **📊 Analysis Questions to Answer :**

**Include short written answers (3–5 sentences) to these:**

* How many rows and columns are in your merged dataset?

* Which pollutants are included, and which have the most missing data?

* How many unique cities or stations are there?

* What are the average levels of key pollutants?

* Are there any immediate data quality issues (e.g., missing or inconsistent values)?

1. **The dataset contains [29,531] ROWS and [16] COLUMNS.**
Refrence : df.shape


2. The dataset includes 12 pollutant columns with their Air Quality Index(AQI) : $\text{PM}_{2.5}$, $\text{PM}_{10}$, $\text{NO}$, $\text{NO}_{2}$, $\text{NO}_\text{x}$, $\text{NH}_3$, $\text{CO}$, $\text{SO}_2$, $\text{O}_3$, Benzene, Toluene, and Xylene.   

The pollutants with the most amount of missing data are:
Xylene: $\mathbf{61.20\%}$ missing values (18,109 out of 29,531 records).$\text{PM}_{10}$: $\mathbf{37.72\%}$ missing values (11,140 out of 29,531 records).$\text{NH}_3$: $\mathbf{34.98\%}$ missing values (10,328 out of 29,531 records).    
Reference :

def missing_values_table(df):

3. **There are 26 unique cities in the dataset.**
Reference : city_files = df['City'].value_counts()

4. The average concentration levels for the most common pollutants are:Average $\mathbf{PM_{2.5}}$: $\mathbf{67.45}$ $\mu g/m^3$Average $\mathbf{PM_{10}}$: $\mathbf{118.12}$ $\mu g/m^3$Average $\text{NO}_\text{x}$: $32.30$ $mg/m^3$

5. Yes, there are two primary and immediate data quality issues:

Significant Missing Data: The most pressing issue is the high percentage of missing values, particularly in the pollutant columns. Xylene is mostly unusable with over $\mathbf{61\%}$ missing, and $\mathbf{PM_{10}}$ and $\mathbf{NH_3}$ also have over $34\%$ missing data.

Incorrect Data Type: The 'Date' column is currently stored as a generic $\mathbf{object}$ (string) data type.